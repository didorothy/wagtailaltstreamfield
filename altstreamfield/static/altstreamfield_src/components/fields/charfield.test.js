import React from "react";
import { render, fireEvent, wait, waitForDomChange, waitForElementToBeRemoved, cleanup } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import CharField from "./charfield";
import FormErrorContext from "../../context/formerror";

describe('CharField', () => {
    test('#minimal', () => {
        let manager = new FormErrorContext.FormErrorManager();
        let container = render(<FormErrorContext.Provider value={manager}>
            <CharField owner_id="struct-01" value="sample text" name="test" onChange={() => {}} />
        </FormErrorContext.Provider>);
        expect(container.container).toMatchSnapshot();
        cleanup();
    });

    test('#label', () => {
        let manager = new FormErrorContext.FormErrorManager();
        let container = render(<FormErrorContext.Provider value={manager}>
            <CharField owner_id="struct-01" value="sample text" name="test" onChange={() => {}} label="Test Char Field" />
        </FormErrorContext.Provider>);
        expect(container.container).toMatchSnapshot();
        cleanup();
    });

    test('#help_text', () => {
        let manager = new FormErrorContext.FormErrorManager();
        let container = render(<FormErrorContext.Provider value={manager}>
            <CharField owner_id="struct-01" value="sample text" name="test" onChange={() => {}} label="Test Char Field" help_text="Sample help text." />
        </FormErrorContext.Provider>);
        expect(container.container).toMatchSnapshot();
        cleanup();
    });

    test('#required', () => {
        let manager = new FormErrorContext.FormErrorManager();
        let container = render(<FormErrorContext.Provider value={manager}>
            <CharField owner_id="struct-01" value="sample text" name="test" onChange={() => {}} label="Test Char Field" help_text="Sample help text." required={true} />
        </FormErrorContext.Provider>);
        expect(container.container).toMatchSnapshot();
        cleanup();
    });

    test('#error', () => {
        let manager = new FormErrorContext.FormErrorManager();
        let container = render(<FormErrorContext.Provider value={manager}>
            <CharField owner_id="struct-01" value="" name="test" onChange={() => {}} label="Test Char Field" help_text="Sample help text." required={true} />
        </FormErrorContext.Provider>);
        expect(container.container).toMatchSnapshot();
        cleanup();
    });

    test('min_length', () => {
        let current_value = "";
        let onChange = (name, value) => {
            current_value = value;
        }
        let manager = new FormErrorContext.FormErrorManager();
        let container = render(<FormErrorContext.Provider value={manager}>
            <CharField owner_id="struct-01" value={current_value} name="test" onChange={onChange} label="Test Char Field" help_text="Sample help text." required={true} min_length={3}/>
        </FormErrorContext.Provider>);
        expect(manager.isValid()).toEqual(false);
        let input = container.container.getElementsByTagName('input')[0];
        let error = container.container.getElementsByClassName('error-message')[0];
        expect(error.textContent).toEqual('This field is required.')

        fireEvent.change(input, {target: {value: 'a'}});
        container.rerender(<FormErrorContext.Provider value={manager}>
            <CharField owner_id="struct-01" value={current_value} name="test" onChange={onChange} label="Test Char Field" help_text="Sample help text." required={true} min_length={3}/>
        </FormErrorContext.Provider>);
        expect(current_value).toEqual('a');
        expect(manager.isValid()).toEqual(false);

        fireEvent.change(input, {target: {value: 'abc'}});
        container.rerender(<FormErrorContext.Provider value={manager}>
            <CharField owner_id="struct-01" value={current_value} name="test" onChange={onChange} label="Test Char Field" help_text="Sample help text." required={true} min_length={3}/>
        </FormErrorContext.Provider>);
        expect(current_value).toEqual('abc');
        expect(manager.isValid()).toEqual(true);
        cleanup();
    });

    test('max_length', () => {
        let current_value = "";
        let onChange = (name, value) => {
            current_value = value;
        }
        let manager = new FormErrorContext.FormErrorManager();
        let container = render(<FormErrorContext.Provider value={manager}>
            <CharField owner_id="struct-01" value={current_value} name="test" onChange={onChange} label="Test Char Field" help_text="Sample help text." required={true} max_length={3}/>
        </FormErrorContext.Provider>);
        expect(manager.isValid()).toEqual(false);
        let input = container.container.getElementsByTagName('input')[0];
        let error = container.container.getElementsByClassName('error-message')[0];
        expect(error.textContent).toEqual('This field is required.')

        fireEvent.change(input, {target: {value: 'a'}});
        container.rerender(<FormErrorContext.Provider value={manager}>
            <CharField owner_id="struct-01" value={current_value} name="test" onChange={onChange} label="Test Char Field" help_text="Sample help text." required={true} max_length={3}/>
        </FormErrorContext.Provider>);
        expect(current_value).toEqual('a');
        expect(manager.isValid()).toEqual(true);

        fireEvent.change(input, {target: {value: 'abcd'}});
        container.rerender(<FormErrorContext.Provider value={manager}>
            <CharField owner_id="struct-01" value={current_value} name="test" onChange={onChange} label="Test Char Field" help_text="Sample help text." required={true} max_length={3}/>
        </FormErrorContext.Provider>);
        expect(current_value).toEqual('abcd');
        expect(manager.isValid()).toEqual(false);
        cleanup();
    });
});