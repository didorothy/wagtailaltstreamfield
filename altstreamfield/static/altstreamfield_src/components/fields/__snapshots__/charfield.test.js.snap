// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharField #error 1`] = `
<div>
  <div
    class="field char_field required error"
  >
    <label
      for="struct-01test"
    >
      Test Char Field
    </label>
    <div
      class="field-content"
    >
      <div
        class="input"
      >
        <input
          id="struct-01test"
          placeholder="Test Char Field"
          type="text"
          value=""
        />
      </div>
      <p
        class="help"
      >
        Sample help text.
      </p>
      <p
        class="error-message"
      >
        This field is required.
      </p>
    </div>
  </div>
</div>
`;

exports[`CharField #help_text 1`] = `
<div>
  <div
    class="field char_field"
  >
    <label
      for="struct-01test"
    >
      Test Char Field
    </label>
    <div
      class="field-content"
    >
      <div
        class="input"
      >
        <input
          id="struct-01test"
          placeholder="Test Char Field"
          type="text"
          value="sample text"
        />
      </div>
      <p
        class="help"
      >
        Sample help text.
      </p>
      
    </div>
  </div>
</div>
`;

exports[`CharField #label 1`] = `
<div>
  <div
    class="field char_field"
  >
    <label
      for="struct-01test"
    >
      Test Char Field
    </label>
    <div
      class="field-content"
    >
      <div
        class="input"
      >
        <input
          id="struct-01test"
          placeholder="Test Char Field"
          type="text"
          value="sample text"
        />
      </div>
      
    </div>
  </div>
</div>
`;

exports[`CharField #minimal 1`] = `
<div>
  <div
    class="field char_field"
  >
    <label
      for="struct-01test"
    >
      Test
    </label>
    <div
      class="field-content"
    >
      <div
        class="input"
      >
        <input
          id="struct-01test"
          placeholder="Test"
          type="text"
          value="sample text"
        />
      </div>
      
    </div>
  </div>
</div>
`;

exports[`CharField #required 1`] = `
<div>
  <div
    class="field char_field required"
  >
    <label
      for="struct-01test"
    >
      Test Char Field
    </label>
    <div
      class="field-content"
    >
      <div
        class="input"
      >
        <input
          id="struct-01test"
          placeholder="Test Char Field"
          type="text"
          value="sample text"
        />
      </div>
      <p
        class="help"
      >
        Sample help text.
      </p>
      
    </div>
  </div>
</div>
`;
